presto-main/src/main/java/io/prestosql/operator/scalar/ConcatWsFunction.java

```diff
@@ -0,0 +1,211 @@
+                        "concat_ws",
+                        ImmutableList.of(),
+                        ImmutableList.of(),
+                        VARCHAR.getTypeSignature(),
+                        ImmutableList.of(VARCHAR.getTypeSignature(), VARCHAR.getTypeSignature()),
+                        true),
+                true,
+                ImmutableList.of(new FunctionArgumentDefinition(false), new FunctionArgumentDefinition(true)),
+                false,
+                true,
+                "Concatenates elements using separator",
+                SCALAR));
+    }
+
+    @Override
+    public ScalarFunctionImplementation specialize(FunctionBinding binding)
+    {
+        if (binding.getArity() < 2) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "There must be two or more arguments");
+        }
+
+        int valueCount = binding.getArity() - 1;
+        MethodHandle arrayMethodHandle = methodHandle(ConcatWsFunction.class, "concatWs", Slice.class, Slice[].class);
+        MethodHandle customMethodHandle = arrayMethodHandle.asCollector(Slice[].class, valueCount);
+
+        return new ChoicesScalarFunctionImplementation(
+                binding,
+                NULLABLE_RETURN,
+                ImmutableList.<InvocationConvention.InvocationArgumentConvention>builder()
+                        .add(NEVER_NULL)
+                        .addAll(Collections.nCopies(valueCount, BOXED_NULLABLE))
+                        .build(),
+                customMethodHandle);
+    }
+
+    @UsedByGeneratedCode
+    public static Slice concatWsArray(Slice separator, Block block)
+    {
+        if (separator == null || block == null) {
+            return null;
+        }
+        Slice[] args = new Slice[block.getPositionCount()];
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            if (block.isNull(i)) {
+                args[i] = null;
+            }
+            else {
+                int sliceLength = block.getSliceLength(i);
+                args[i] = block.getSlice(i, 0, sliceLength);
+            }
+        }
+        return concatWs(separator, args);
+    }
+
+    @UsedByGeneratedCode
+    public static Slice concatWs(Slice separator, Slice[] values)
+    {
+        if (separator == null || values == null) {
+            return null;
+        }
+
+        // Validate size of output
+        int length = 0;
+        boolean requiresSeparator = false;
+
+        for (Slice value : values) {
+            if (value == null) {
+                continue;
+            }
+            try {
+                if (requiresSeparator) {
+                    length = addExact(length, separator.length());
+                }
+                length = addExact(length, value.length());
+                requiresSeparator = true;
+            }
+            catch (ArithmeticException e) {
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Concatenated string is too large");
```

Comment:
This will happen when the length is larger than Integer.MAX_VALUE. This is way too big. The engine can't handle pages that big anyway. I'd cap this to a much lower value, in the order of megabytes (current target size for pages is ~1MB -- see `PageBuilderStatus.DEFAULT_MAX_PAGE_SIZE_IN_BYTES`).

Reply:
I've added a stricter upper bound, and removed the ArithmeticException check for brevity, which is very unlikely to reach given this tighter bound.